python/paddle/fluid/dygraph/dygraph_to_static/loop_transformer.py

```diff
@@ -0,0 +1,251 @@
+        gast.Name(
+            id=body_name, ctx=gast.Param(), annotation=None, type_comment=None))
+    assign_targets = [
+        gast.Name(
+            id=var_name, ctx=gast.Param(), annotation=None, type_comment=None)
+        for var_name in loop_var_names
+    ]
+    while_args.append(gast.List(elts=assign_targets, ctx=gast.Param()))
+
+    while_func_id = gast.parse('fluid.layers.while_loop').body[0].value
+    while_node = gast.Call(func=while_func_id, args=while_args, keywords=[])
+    assign_node = gast.Assign(
+        targets=[gast.Tuple(
+            elts=assign_targets, ctx=gast.Store())],
+        value=while_node)
+    return assign_node
+
+
+class NameVisitor(gast.NodeVisitor):
+    '''
+    Analysis name liveness for loop transformer
+    '''
+
+    def __init__(self, root_node):
+        # Set of gast.Name
+        self.current_seen_vars = set()
+        # List of gast.While/gast.For nodes
+        self.current_loop = []
+
+        # Mapping from gast.While/gast.For to string name of vars
+        self.before_loop_vars = defaultdict(set)
+        self.in_loop_vars = defaultdict(set)
+
+        self.visit(root_node)
+
+    def is_control_flow_loop(self, node):
+        # TODO: make a better condition
+        return True
+
+    def get_loop_var_names(self, node):
+        assert isinstance(node, gast.While) or isinstance(
+            while_node, gast.For), "Input node is not gast loop node"
+        loop_var_names = set()
+        create_var_names = set()
+        read_context = {type(gast.Load), type(gast.AugLoad)}
+
+        in_loop_vars = self.in_loop_vars[node]
+        in_loop_name_strs = set(name.id for name in in_loop_vars)
+        before_loop_vars = self.before_loop_vars[node]
+        before_loop_name_strs = set(name.id for name in before_loop_vars)
+        after_loop_vars = self.current_seen_vars - before_loop_vars - in_loop_vars
+        after_loop_name_strs = set(
+            name.id for name in after_loop_vars
+            if type(name.ctx) in read_context)
+        for name in in_loop_name_strs:
+            if name in before_loop_name_strs:
+                # If a variable is used in loop and created before loop, it
+                # should be in loop_var as input
+                loop_var_names.add(name)
+            elif name in after_loop_name_strs:
+                # If a variable is created in the while loop and read after
+                # loop, it should be in loop_var and we should create it
+                loop_var_names.add(name)
+                create_var_names.add(name)
+        return loop_var_names, create_var_names
+
+    def visit_Name(self, node):
+        self.current_seen_vars.add(node)
+        for loop_node in self.current_loop:
+            self.in_loop_vars[loop_node].add(node)
+        self.generic_visit(node)
+
+    def visit_For(self, node):
+        self.current_loop.append(node)
+        self.before_loop_vars[node] = copy.deepcopy(self.current_seen_vars)
```

Comment:
There are known bugs in these "copy.deepcopy", but due to slow CI, I prefer to change them in

https://github.com/PaddlePaddle/Paddle/pull/22847